src/Backups/BackupIO_AzureBlobStorage.cpp

```diff
@@ -0,0 +1,329 @@
+    client = object_storage->getClient();
+}
+
+BackupReaderAzureBlobStorage::~BackupReaderAzureBlobStorage() = default;
+
+bool BackupReaderAzureBlobStorage::fileExists(const String & file_name)
+{
+    String key;
+    if (startsWith(file_name, "."))
+    {
+        key= configuration.blob_path + file_name;
+    }
+    else
+    {
+        key = file_name;
+    }
+    return object_storage->exists(StoredObject(key));
+}
+
+UInt64 BackupReaderAzureBlobStorage::getFileSize(const String & file_name)
+{
+    String key;
+    if (startsWith(file_name, "."))
+    {
+        key= configuration.blob_path + file_name;
+    }
+    else
+    {
+        key = file_name;
+    }
+    ObjectMetadata object_metadata = object_storage->getObjectMetadata(key);
+    return object_metadata.size_bytes;
+}
+
+std::unique_ptr<SeekableReadBuffer> BackupReaderAzureBlobStorage::readFile(const String & file_name)
+{
+    String key;
+    if (startsWith(file_name, "."))
+    {
+        key= configuration.blob_path + file_name;
+    }
+    else
+    {
+        key = file_name;
+    }
+    return std::make_unique<ReadBufferFromAzureBlobStorage>(
+        client.get(), key, read_settings, settings->max_single_read_retries,
+        settings->max_single_download_retries);
+}
+
+void BackupReaderAzureBlobStorage::copyFileToDisk(const String & path_in_backup, size_t file_size, bool encrypted_in_backup,
+                                    DiskPtr destination_disk, const String & destination_path, WriteMode write_mode)
+{
+    LOG_INFO(&Poco::Logger::get("BackupReaderAzureBlobStorage"), "Enter copyFileToDisk");
+
+    /// Use the native copy as a more optimal way to copy a file from AzureBlobStorage to AzureBlobStorage if it's possible.
+    /// We don't check for `has_throttling` here because the native copy almost doesn't use network.
+    auto destination_data_source_description = destination_disk->getDataSourceDescription();
+    if (destination_data_source_description.sameKind(data_source_description)
+        && (destination_data_source_description.is_encrypted == encrypted_in_backup))
+    {
+        LOG_TRACE(log, "Copying {} from AzureBlobStorage to disk {}", path_in_backup, destination_disk->getName());
+        auto write_blob_function = [&](const Strings & blob_path, WriteMode mode, const std::optional<ObjectAttributes> & object_attributes) -> size_t
+        {
+            /// Object storage always uses mode `Rewrite` because it simulates append using metadata and different files.
+            if (blob_path.size() != 2 || mode != WriteMode::Rewrite)
+                throw Exception(ErrorCodes::LOGICAL_ERROR,
+                                "Blob writing function called with unexpected blob_path.size={} or mode={}",
+                                blob_path.size(), mode);
+
+            copyAzureBlobStorageFile(
+                client,
+                reinterpret_cast<AzureObjectStorage *>(destination_disk->getObjectStorage().get())->getClient(),
```

Comment:
Sometimes an object storage wraps around another object storage, for example `CachedObjectStorage`. It seems this `reinterpret_cast` will not work in that case.

Reply:
Ah okay, then we cannot get the Client from the Disk. I will revert back to previous method of establing connection if destintaion container/source container are different. 
Let me know if you have any other suggestions of getting client

Reply:
Why is that? Yes we can get the client from the Disk.

Just make the function `getClient()` virtual:
```
namespace Azure::Storage::Blobs { class BlobContainerClient; }

class IObjectStorage
{
   virtual std::shared_ptr<Azure::Storage::Blobs::BlobContainerClient> getAzureBlobStorageClient() const
   {
        throw Exception(NOT_IMPLEMENTED ...);
   }
};

class AzureObjectStorage : public IObjectStorage
{
    std::shared_ptr<Azure::Storage::Blobs::BlobContainerClient> getAzureBlobStorageClient() const override
    {
        return client;
    }
};

class CachedObjectStorage : public IObjectStorage
{
    std::shared_ptr<Azure::Storage::Blobs::BlobContainerClient> getAzureBlobStorageClient() const override
    {
        return object_storage->getAzureBlobStorageClient();
    }
}
```

Reply:
ya we can do that. I just feel `getAzureBlobStorageClient` is quite specific function to put in IObjectStorage interface, maybe we can call it `getClient()` & keep it not implemented in ObjectStorage by default except for AzureObjectStorage.  Maybe we can extend this to other disk too. WDYT ?

Reply:
Yes, it's quite specific, but it's not a problem. For any other storage type it will just throw an exception. It doesn't make much sense in a general name like `getClient()` because this function is going to return `std::shared_ptr<Azure::Storage::Blobs::BlobContainerClient>` anyway.

Reply:
addressed it